lib/boundkeypair/bound_keypair.go

```diff
@@ -0,0 +1,176 @@
+
+		// TODO: API design issue to consider when implementing rotation: the
+		// public key will change during rotation. Should a new validator be
+		// created, or can we design this better?
+		publicKey: publicKey,
+	}, nil
+}
+
+func (v *ChallengeValidator) IssueChallenge() (*ChallengeDocument, error) {
+	// Implementation note: these challenges are only ever sent to a single
+	// client once, and we expect a valid reply as the next exchange in the
+	// join ceremony. There is not an opportunity for reuse, multiple attempts,
+	// or for clients to select their own nonce, so we won't bother storing it.
+	nonce, err := utils.CryptoRandomHex(defaults.TokenLenBytes)
+	if err != nil {
+		return nil, trace.Wrap(err, "generating nonce")
+	}
+
+	return &ChallengeDocument{
+		Claims: &jwt.Claims{
+			Issuer:    v.clusterName,
+			Audience:  jwt.Audience{v.clusterName}, // the cluster is both the issuer and audience
+			NotBefore: jwt.NewNumericDate(v.clock.Now().Add(challengeNotBeforeOffset)),
+			IssuedAt:  jwt.NewNumericDate(v.clock.Now()),
+			Expiry:    jwt.NewNumericDate(v.clock.Now().Add(challengeExpiration)),
+			Subject:   v.subject,
+		},
+		Nonce: nonce,
+	}, nil
+}
+
+// ValidateChallengeResponse validates a signed challenge document, ensuring the
+// signature matches the requested public key, and that the claims pass JWT
+// validation.
+func (v *ChallengeValidator) ValidateChallengeResponse(issued *ChallengeDocument, compactResponse string) error {
+	token, err := jwt.ParseSigned(compactResponse)
+	if err != nil {
+		return trace.Wrap(err, "parsing signed response")
+	}
+
+	var document ChallengeDocument
+	if err := token.Claims(v.publicKey, &document); err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Validate the challenge document claims per JWT rules.
+	const leeway time.Duration = time.Minute
+	if err := document.Claims.ValidateWithLeeway(jwt.Expected{
+		Issuer:   v.clusterName,
+		Subject:  v.subject,
+		Audience: jwt.Audience{v.clusterName},
+		Time:     v.clock.Now(),
+	}, leeway); err != nil {
+		return trace.Wrap(err, "validating challenge claims")
+	}
+
+	// JWT validation won't check equality on the time fields, so we'll manually
+	// check them.
+	if !issued.Claims.IssuedAt.Time().Equal(document.Claims.IssuedAt.Time()) {
+		return trace.AccessDenied("invalid challenge document")
+	}
+	if !issued.Claims.Expiry.Time().Equal(document.Claims.Expiry.Time()) {
+		return trace.AccessDenied("invalid challenge document")
+	}
+	if !issued.Claims.NotBefore.Time().Equal(document.Claims.NotBefore.Time()) {
+		return trace.AccessDenied("invalid challenge document")
+	}
+
+	if subtle.ConstantTimeCompare([]byte(issued.Nonce), []byte(document.Nonce)) == 0 {
+		return trace.AccessDenied("invalid nonce")
+	}
+
+	return nil
+}
+
+// RecoveryModes returns a list of all supported recovery modes
+func RecoveryModes() []RecoveryMode {
+	return []RecoveryMode{
+		RecoveryModeStandard,
+		RecoveryModeRelaxed,
+		RecoveryModeInsecure,
+	}
+}
```

Comment:
Consider moving this block up next to where the various modes are declared so it's easier to keep them in sync.